### DISCUSSION TITLE
ManagedUI Installer does not create Uninstall registry keys?

### DISCUSSION THREAD

**Discussion by dbieber-rfideas** [oleg-shilo/wixsharp#1617] [Q&A]
Hello,

The application I am migrating to wix (wixsharp) is relying on searching the Uninstall key of the registry at this location to determine its own product code for business logic it performs:
`HKEY_LOCAL_MACHINE\SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\`

The old installer (Microsoft Visual Studio Installer Projects) created entries in this key, but Wix does not.
I've tried explicitly adding an Uninstall UI because the WixSharp source code seems to add registry keys for that operation, but I suspect it ignores this since the key doesn't exist in the first place:

```csharp
        /// <summary>
        /// Internal UninstallFullUI action. It must be public for the DTF accessibility but it is not to be used by the user/developer.
        /// </summary>
        /// <param name="session">The session.</param>
        /// <returns></returns>
        [CustomAction]
        public static ActionResult WixSharp_EnableUninstallFullUI_Action(Session session)
        {
            return session.HandleErrors(() =>
            {
                var productCode = session.Property("ProductCode");
                var extraParameters = session.Property("MSIEXEC_EXTRA_PARAMETERS");

                var keyName = $@"SOFTWARE\Microsoft\Windows\CurrentVersion\Uninstall\{productCode}";

                using (var uninstallKey = Registry.LocalMachine.OpenSubKey(keyName, true))
                {
                    if (uninstallKey == null)
                    {
                        return;
                    }

                    uninstallKey.SetValue("ModifyPath", string.Empty);
                    uninstallKey.SetValue("UninstallString", $"MsiExec.exe /I{productCode} {extraParameters}");
                    uninstallKey.SetValue("WindowsInstaller", 0);
                }
            });
        }
```

I assume `uninstallKey` ends up being null and that this is not doing me any good.

Here's a snippet of my wixsharp script:

```csharp
        // Build a WixSharp project
        var project = new ManagedProject(Options["ProductName"],
            new Dir($@"%ProgramFiles%\{config.Manufacturer}\{Options["ProductName"]}",
                new File(new Id("application_exe"), binaries,
                    $@"{Options["ArtifactPublishDirectoryPath"]}/{Options["ProductName"]}.exe",
                    new FileShortcut(binaries, Options["ProductName"], "%ProgramMenu%"),
                    new FileShortcut(binaries, Options["ProductName"], "%Desktop%")
                ),
                new File(binaries, config.ProductIconFile.FullName),
                new Files(binaries, $"{Options["ArtifactPublishDirectoryPath"]}/*.*",
                    f => !f.EndsWith(".pdb") && !f.EndsWith($"{Options["ProductName"]}.exe"))
            ),
            new Binary(new Id("License"), config.LicenseFile.FullName)
        )
        {
            GUID = config.ProductCode,
            UpgradeCode = config.UpgradeCode,
            ManagedUI = new ManagedUI()
            {
                InstallDialogs = new ManagedDialogs()
                    .Add(Dialogs.Welcome)
                    .Add(Dialogs.Licence)
                    // .Add(Dialogs.Features)
                    // .Add(Dialogs.SetupType)
                    .Add(Dialogs.InstallDir)
                    .Add(Dialogs.Progress)
                    .Add(Dialogs.Exit),
                ModifyDialogs = new ManagedDialogs()
                    .Add(Dialogs.MaintenanceType)
                    .Add(Dialogs.Progress)
                    .Add(Dialogs.Exit)
            },

            Version = new Version(Options["Version"]),
            InstallerVersion = 301,
            BannerImage = config.BannerGraphicFile.FullName,
            BackgroundImage = config.PanelGraphicFile.FullName,
            LicenceFile = config.LicenseFile.FullName,
            ControlPanelInfo = new ProductInfo
            {
                Manufacturer = config.Manufacturer,
                ProductIcon = config.ProductIconFile.FullName,
                Comments = config.Description,
                Contact = config.Contact,
                HelpLink = config.HelpLink,
                UrlInfoAbout = config.UrlInfoAbout,
                UrlUpdateInfo = config.UrlUpdateInfo
            },
            Platform = Platform.x86,
            MajorUpgradeStrategy = new MajorUpgradeStrategy()
            {
                UpgradeVersions = VersionRange.ThisAndOlder,
                RemoveExistingProductAfter = Step.InstallInitialize,
                PreventDowngradingVersions = VersionRange.NewerThanThis,
                NewerProductInstalledErrorMessage =
                    "A later version of [ProductName] is already installed. Setup will now exit."
            },
        };

        project.EnableUninstallFullUI();
```

I've also looked in the WOW6432Node/ hive, and I've looked in HKCU. In fact, I searched the whole registry for my application name and for its product code and nothing comes up, except a couple of WixSharp Used/ entries.

Examining the wix XML output, I see no indication that any registry keys would be created for Uninstall.

What am I missing?

**Comment by oleg-shilo** [comment]
You might be having a permissions problem. By default your CA is _immediate_ so it is not elevated. So it may just fail writing to the registry. 

If it is the case then you can use managed event `AfterInstall`. It is already elevated

**Comment by dbieber-rfideas** [comment]
Thank you oleg.

I am not actually writing to the registry myself - the question was about registry keys that should be automatically created by the installer framework (registering the uninstaller, I guess?).

I'm not even sure how to create the registry keys that it is looking for. All I can say is, Visual Studio Installer Projects (.vdproj files) do this for you, but I do not know why. Our developers ended up relying on that information being in the registry, which is no longer the case after switching to Wix/WixSharp.

That being said, this is a low priority for me. I ended up just modifying our application code to remove the dependence on the registry. In this case, there was another way we could solve our problem, which lets us dodge this issue for now.

If we ever have to add custom registry keys in the future, I will try to keep what you said in mind. However, I think the behavior is the same when I simply run the installer as an administrator (not usually allowed, but we can for testing). I'm not 100% sure we performed this troubleshooting step, so take that with a grain of salt.

**Comment by oleg-shilo** [comment]
The msi you are building does not write to the the registry unless you are doing it in the custom action. It is MSI runtime that creates hips of registry values as part of the product installation. It does it regardless of which tool you use to create the msi.

Though, quite possibly InstallScope (`project.Scope`) may give you some control over it.
